mixin for intervalmetadata

RNAer opened the pull request:
This is a followup PR of #1414 .

@jairideout and @ebolyen , could you take a quick look to see if the mixin class API and the way how the repr represents the object are ok? don't worry about the tests. I will add more docstring and fix tests.

The next task is to modify all the sequence classes to adapt the addition of `IntervalMetadata`.

The next next task is to modify GenBank parser.

RNAer commented:
@wasade , @josenavas, @ekopylova, it would be great if you can take a look too.

mortonjt commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+
+        >>> from skbio import DNA
```
Need to have `Examples` header

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+
+
+class IntervalMetadataMixin(metaclass=abc.ABCMeta):
+    def __init__(self, interval_metadata=None):
```
Decorate with `@abc.abstractmethod`

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+    def __init__(self, interval_metadata=None):
+        raise NotImplementedError
+
+    def _init_(self, interval_metadata=None):
```
Please structure methods in this class like they are in the positional metadata mixin. I'm trying to keep the 3 mixin classes as consistent as possible so it's easier to maintain and reason about them, and certain design decisions were made for safety and performance. It'll also make writing the tests easier because you can model those off of `metadata` or `positional_metadata`. For example, `_init_` should be:

```python
def _init_(self, interval_metadata=None):
    if interval_metadata is None:
        # Could use deleter but this is less overhead and needs to be fast.
        self._interval_metadata = None
    else:
        # Use setter for validation and copy.
        self.interval_metadata = interval_metadata
```

I'll mark any other methods that need updating to match positional metadata mixin.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata()
```
We'll need the `IntervalMetadata` constructor to accept a length, and validate that interval features all fall within that coordinate space when they are added. This is necessary to ensure that interval metadata spans the length of the sequence, similar to how the number of rows in positional metadata (`pd.DataFrame`) match the sequence length.

This doesn't have to be called "length", since that could be confused with the number of interval features in the interval metadata. I'm not sure what to call it, do you have any ideas?

Once `IntervalMetadata` has a "length", `IntervalMetadata._reverse` can be renamed `.reverse()` and will no longer need to accept `length` as a parameter (since it will be instance state).

This mixin will need an abstract method similar to `_positional_metadata_axis_len_` that objects must implement to define their length. Probably makes sense to call it `_interval_metadata_axis_len_` (or some variation if we have a better name than "length").

RNAer replied:
This sounds good to me. how about `upper_bound` or `sequence_length` for the nameing?

jairideout replied:
I like `upper_bound`. It seems interval metadata *is* an interval from `[0, N]`, where `N` is the `upper_bound`, and interval metadata contains regions (contiguous or non-contiguous) that are identified to have particular metadata attached to them.

`sequence_length` would make sense if interval metadata was only attached to sequences, but it can be applied to any other scikit-bio object that has a notion of length (e.g. `TabularMSA`).

@ebolyen do you have thoughts about naming here?

ebolyen replied:
`_interval_metadata_axis_len_` makes sense to me... We aren't saying anything about the interval metadata so much as the limitation of the axis it exists on (as defined by the mixin-ee/mixee).

jairideout replied:
This discussion is about what the name `IntervalMetadata` constructor's parameter defining its "length".

ebolyen replied:
Sorry I misunderstood. What about `axis_length`? then it matches the "sunder" method and doesn't imply too much.

jairideout replied:
"axis length" makes sense for the object this mixin is applied to, since some objects have more than one axis (e.g. for `TabularMSA`, it could be attached to rows or columns. `Sequence` is less interesting.). For `IntervalMetadata` though, it really is a closed interval `[0, N]`. `N` is what we're trying to name here (which will be the constructor parameter). Using "axis" in `IntervalMetadata` doesn't seem as clear to me, that's why I was thinking something like the `upper_bound` of the interval would make more sense. Does that distinction make sense, or is "axis" really applicable to `IntervalMetadata` standalone? Note that this is similar to how positional metadata mixin defines `_positional_metadata_axis_len_`, and the `pd.DataFrame`'s "length" is the number of rows (the two don't directly correspond, since a data frame has two axes).

ebolyen replied:
I suppose the name does kind of have "interval" in it, so `upper_bound` does make sense. But I think it can also be said that it *has* a continuous axis (which does not necessarily need to be bounded, but is in practice). I think any name has the potential to look weird in the context of how we decide to implement `iter` and friends, which is of course unknown.

I think I see this object as closer to a line with references on it, which makes me intuitively feel like the line comes first and the features second (i.e. not a mapping) which is why I like `axis_length`, and `length` feels even more natural. **BUT** that assumes something really important about the object (and makes `__len__` pretty pointless), so I'll defer to you on naming.

RNAer replied:
ok. Let's settle on `upper_bound` then.

jairideout replied:
Thanks for your input @ebolyen, that makes a lot of sense. The parameter name seems to depend on how we're conceptually thinking of the `IntervalMetadata`'s coordinate space (bounded vs unbounded). Given that I'm fine with either name. Let's go with `upper_bound`, @RNAer is the tie-breaker :)

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+
+    @interval_metadata.setter
+    def interval_metadata(self, interval_metadata):
+        if interval_metadata is None:
```
Remove support for `None` since the constructor, getter, and deleter will handle this (see positional metadata's setter or metadata's setter for examples).

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+            interval_metadata = IntervalMetadata()
+        if isinstance(interval_metadata, IntervalMetadata):
+            # copy all the data to the mixin
+            self._interval_metadata = copy.deepcopy(interval_metadata)
```
We want a shallow copy here instead of a deep copy to match the other metadata mixins.

`IntervalMetadata` doesn't define a `__copy__` method, which is what `copy.copy` calls (`copy.copy` performs a shallow copy). It *may* work as-is, but can you please add an `IntervalMetadata.__copy__` method with those semantics defined explicitly? I tend to not trust that custom classes will work with `copy.copy` (for example, numpy arrays and pandas dataframes don't always work with `copy.copy`, though that may have been fixed in more recent versions).

Also, above this line is where we'd want to validate that the supplied `interval_metadata` object's length matches `self._interval_metadata_axis_len_()`. This validation is similar to the positional metadata setter.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+    def has_interval_metadata(self):
+        """Determine if the object has interval metadata.
+
+        An object has interval metadata if its ``interval_metadata``
```
👍

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+        raise NotImplementedError
+
+    def _eq_(self, other):
+        return self.interval_metadata == other.interval_metadata
```
See the logic for positional metadata mixin's `_eq_` method -- there's more complicated logic to handle cases where positional metadata is missing.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+
+    def _copy_(self):
+        if self.has_interval_metadata():
+            return copy.copy(self.interval_metadata)
```
See [my comment](https://github.com/biocore/scikit-bio/pull/1423/files#r77926433) about defining `IntervalMetadata.__copy__`. This code won't need updating because `copy.copy` will handle calling the magic method.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+
+    def _deepcopy_(self, memo):
+        if self.has_interval_metadata():
+            return copy.deepcopy(self.interval_metadata, memo)
```
Related to my comment [here](https://github.com/biocore/scikit-bio/pull/1423/files#r77926433). Can you define `IntervalMetadata.__deepcopy__` to ensure this works as we intend? This code won't need updating because `copy.deepcopy` will handle calling the magic method once it exists.

jairideout commented on skbio/metadata/_repr.py:
```diff
@@ -126,6 +127,13 @@ def _format_positional_metadata_column(self, key, dtype):
+        if self._obj.has_interval_metadata():
+            self._lines.add_line('Interval metadata:')
+            self._lines.add_line(
+                    self._indent + '%d interval features' % len(
```
Can you handle the singular case? i.e. "1 interval feature"

RNAer replied:
sure. so the seq repr looks ok to show just the number of interval features as in line 430?

jairideout replied:
Yeah I think the repr looks great!

jairideout commented:
@RNAer overall this looks great! Most of my comments are related to matching the behavior of the other metadata mixin classes, and defining additional methods on `IntervalMetadata` that the mixin will use. I also had some comments about validation, which requires `IntervalMetadata` to have a length. The repr looks good.

I didn't look at the tests, will do that when you're ready and these larger comments have been addressed. Please let me know if you have any questions about my review comments.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,125 @@ def has_positional_metadata(self):
+        True
+        """
+        return (self._interval_metadata is not None and
+                len(self.interval_metadata._intervals) > 0)
```
Maybe it makes sense for `IntervalMetadata` to have a read-only property, something like `num_intervals` or `num_interval_features` so that this mixin doesn't rely on private API? This would avoid the repr builder from accessing that private state too.

RNAer replied:
what about implementing `__len__` and `__iter__`? this is more pythonic?

jairideout replied:
It'd be more Pythonic, but we had originally deferred defining `__len__`, `__iter__`, `__getitem__`, `__setitem__`, etc., which all must agree on what "length" means for `IntervalMetadata`. We weren't able to come to a conclusion in the RFC because we weren't sure how/if `Sequence.__getitem__` would use `IntervalMetadata.__getitem__`. `__getitem__`'s behavior will determine the behavior of `__len__`, `__iter__`, etc. so we can't define those yet.

Defining these methods depends on whether we see `IntervalMetadata` as a Python "sequence" or a "mapping". For details on sequence and mapping protocols, see [[1]](https://docs.python.org/3/reference/datamodel.html#emulating-container-types) and [[2]](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). I think we should defer defining these methods until we have a clearer idea of what makes the most sense / what is useful for them to do (and that'll require some discussion). To move this forward, I think we should define an attribute with a clear name that isn't tied to the magic methods listed above. That way we are free to define those later in a backwards-compatible way. I like `num_interval_features` or `interval_feature_count` since they're very descriptive names, and that's how they're described in the repr and `has_interval_metadata`. What do you think?

jairideout replied:
@ebolyen do you have thoughts on this?

ebolyen replied:
:+1: for delaying the builtins, I have no strong feelings on the name though.

jairideout replied:
Thanks @ebolyen! @RNAer, I'm leaning towards `num_interval_features` or `interval_feature_count`. Do you think either of those names makes sense here?

RNAer replied:
I was thinking only defining `__len__` and `__iter__`, not `__getitem__` and `__setitem__`. But what you said makes sense. Let's defer it. I will choose `num_interval_features` because it is one char shorter :)

jairideout replied:
Sounds good, thanks!

RNAer commented:
Thanks, @jairideout . I have replied your 3 comments and if you think ok, I can add those to `IntervalMetadata` in another PR.

jairideout commented:
Thanks @RNAer! I replied to your comments, also soliciting feedback from @ebolyen. Let's finalize those discussions and then it'd be great if you could make those `IntervalMetadata` changes in another PR that is merged before this one. Thanks!

jairideout commented:
@RNAer, I think we've settled on the API changes to `IntervalMetadata`. I'll be on the lookout for that PR, the review should be pretty quick.

RNAer commented:
@mortonjt , could you take a pass thru this PR if you are bored at east coast? : )

@jairideout , this is ready for review, the test failed because the various places calls `has_interval_metadata` in `__repr__` in `Sequence` and its child classes. Adding interval metadata mixin to all the classes is gonna be a big PR. do you think it is ok? or there is good way to decompose it?

mortonjt commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,166 @@ def test_has_positional_metadata(self):
+            # repeat it
+            {'locations': [(0, 1)], 'metadata': {'gene': ['a'],
+                                                 'product': 'foo'}}]
+
```
Would it be worthwhile to have a different interval other than (0, 1) or multiple intervals? It would be nice to double check if the equality works as anticipated.

mortonjt commented:
Concerning the failing tests,  I wouldn't merge this branch into master, but merge it into another branch,  so that the changes for Sequence and other subclasses can be reviewed separately

RNAer commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,157 @@ def has_positional_metadata(self):
+    def _interval_metadata_axis_len_(self):
+        ''''''
+        raise NotImplementedError
+
```
pls c.f. `PositionalMetadataMixin`. basically the child class inheriting the mixins handle that in its own polymorphismi way

RNAer commented:
Thanks, @mortonjt . @jairideout , do you think it is a good idea to PR agains a non-master branch, review code there, accumulate changes there, and merge all to master from there?

jairideout commented:
@RNAer, I like how you've been splitting up these PRs into smaller, atomic chunks that get merged in -- it helps a lot with reviewing and keeps master branch moving forward with incremental improvements. I'd prefer to keep using this strategy of merging smaller PRs into master branch when it's possible. If not, we can do what @mortonjt is suggesting by using a separate feature branch.

I'll follow up with an inline comment showing how we can fix the Sequence/TabularMSA repr issue so that those classes can be updated in a subsequent PR (it's an easy fix in this case).

jairideout commented on skbio/metadata/_repr.py:
```diff
@@ -126,6 +127,15 @@ def _format_positional_metadata_column(self, key, dtype):
         return self._wrap_text_with_indent(dtype_fmt, key_fmt, 1)
 
+    def _process_interval_metadata(self):
+        if self._obj.has_interval_metadata():
```
To work around the Sequence/TabularMSA repr issues, add this:

```python
if hasattr(self._obj, "has_interval_metadata") and self._obj.has_interval_metadata():
    ...
```

This is a little gross, but this repr class (`_MetadataReprBuilder`) already needs some refactoring so that classes can mix and match which types of metadata they have and need reprs for (`.metadata`, `.positional_metadata`, `.interval_metadata`). It probably makes sense to remove this class and move each of the metadata repr helpers into their respective metadata mixin classes (e.g. as `_repr_` methods). This is a larger change that's not super important to make (it's all private API) so I think using the `hasattr` workaround is a practical way to move forward here.

It's also fine if you want to remove the repr updates (and mixin doctests) from this PR, and add them in the upcoming Sequence PR. I think we'll still need the `hasattr` workaround.

RNAer commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,166 @@ def test_has_positional_metadata(self):
+        im2.add(**self.intvls[0])
+        obj2 = self._interval_metadata_constructor_(self.upper_bound, im2)
+
+        self.assertReallyEqual(obj1, obj2)
```
@jairideout , I think we can ditch `skbio.util.ReallyEqualMixin` here, right? I read the webpage linked in its docstring (https://ludios.org/testing-your-eq-ne-cmp/). Basically, the justification for this is gone in Python 3. `__cmp__` is deprecated and the fallback of `__ne__` returns the negated result of` __eq__` in Python3.

jairideout replied:
We should continue using `skbio.util._testing.ReallyEqualMixin` because it performs other equality sanity checks, such as making sure `A == B` and `B == A`, and `A != B` and `B != A`. This can catch subtle bugs in equality implementations. The mixin in scikit-bio doesn't include the `cmp` checks as those are no longer necessary in Python 3.

jairideout commented:
👍 for adding reverse complement tests. Also I noticed coverage dropped, can you run coverage locally and add tests as needed? I'll take a pass thru when you're done with that.

RNAer commented:
The reverse method is under `IntervalMetadata` not `IntervalMetadataMixin` - why adding a test here?

mortonjt commented:
This is mainly to make sure that `_axis_len` is being interacted with properly, since there is a significant amount of length tinkering in this PR.  Otherwise, I don't see any additional issues (outside of @jairideout 's comments) that need to be addressed.

jairideout commented:
I agree, 👍 for some basic reverse tests to make sure length is being passed thru correctly

RNAer commented:
While I added the reverse test, I still dont see this is needed. the reverse() is isolated in `IntervalMetadata` and there is no variable passing between it and anything else. Anyway.

@jairideout , the test coverage drop is mainly due to the remove of repr doctests. Now I removed repr modification and will add it back in next PR where the mixin will be added to `Sequence`. I think this is ready for you to take a pass.

jairideout commented:
Great, thanks @RNAer! I'll review tomorrow morning.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
```
Can you add a `Notes` section, similar to metadata/positional metadata? Something like:

```
Notes
-----
This property can be set and deleted. When setting new interval metadata, a
shallow copy of the `IntervalMetadata` object is made.
```

We'll need an `Examples` section too but will defer that until this class is hooked up to `Sequence` in another PR.

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
+            # copy all the data to the mixin
+            self._interval_metadata = copy.copy(interval_metadata)
+        else:
+            raise TypeError('You must provide `IntervalMetadata` object.')
```
More informative error message to help users, something like:

```python
raise TypeError("Interval metadata must be an `IntervalMetadata` object, not type %r" % type(interval_metadata).__name__)
```

jairideout commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
+        if isinstance(interval_metadata, IntervalMetadata):
+            upper_bound = interval_metadata.upper_bound
+            axis_len = self._interval_metadata_axis_len_()
+            if upper_bound != axis_len:
```
Can you also check that `.lower_bound` is zero? Currently it will always be zero but that may not be true in the future, and we'd want to catch that here.

jairideout replied:
There is an off-by-one error here. IntervalMetadata lower/upper bounds are inclusive, so this code forces the IntervalMetadata length to be one position greater than the object's axis length. For example, if you have a Sequence with 4 bases, it's axis length will be 4, so the IntervalMetadata's bounds should be `[0, 3]`. The check should be:

```python
if upper_bound != (axis_len - 1):
    ...
```

Can you add unit tests that would have caught this bug?

RNAer replied:
@jairideout , the current code is correct. The bounds are working like Python index. The `[(0, 4)]` means nucleotide from 0 to 3, not including 4. and the `upper_bound` should be 4 for a sequence of 4.

jairideout replied:
I am pretty certain there is an off-by-one error. IntervalMetadata lower/upper bounds are inclusive, so having a lower bound of 0 and an upper bound of 4 (in this example) is five positions, where the number of positions in the Sequence is 4. Here is an example using the `ExampleIntervalMetadataMixin` test object:

```python
In [1]: from skbio.metadata._mixin import IntervalMetadataMixin
In [2]: from skbio.metadata import IntervalMetadata
In [3]: from skbio.util._decorator import overrides
In [4]: # copy and paste the ExampleIntervalMetadataMixin class definition
        # from the unit tests, not included here for brevity

# Create an object of length 4 (i.e. 4 positions). The current code requires
# its IntervalMetadata to have an upper bound of 4, which is inclusive.
In [5]: e = ExampleIntervalMetadataMixin(4, IntervalMetadata(4))

# Confirm the object's axis length is 4 (i.e. 4 positions).
In [6]: e._interval_metadata_axis_len_()
Out[6]: 4

# The IntervalMetadata's lower bound is 0 (inclusive).
In [7]: e.interval_metadata.lower_bound
Out[7]: 0

# The IntervalMetadata's upper bound is 4 (inclusive).
In [8]: e.interval_metadata.upper_bound
Out[8]: 4

# Confirm IntervalMetadata's bounds are inclusive by adding metadata
# that includes the lower and upper bounds (0 and 4, respectively).
In [9]: e.interval_metadata.add(locations=[(0, 4)], metadata={'foo': 42})
Out[9]: Interval(interval_metadata=<4592643040>, locations=[(0, 4)], boundaries=[(True, True)], metadata={'foo': 42})

# Confirm that IntervalMetadata disallows an interval with an upper bound
# of 5.
In [10]: e.interval_metadata.add(locations=[(0, 5)], metadata={'bar': 43})
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-2867edfcc7cf> in <module>()
----> 1 e.interval_metadata.add(locations=[(0, 5)], metadata={'bar': 43})

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in add(self, locations, boundaries, metadata)
    588                         locations=locations,
    589                         boundaries=boundaries,
--> 590                         metadata=metadata)
    591
    592     @_rebuild_tree

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in __init__(self, interval_metadata, locations, boundaries, metadata)
     80         self._interval_metadata = interval_metadata
     81
---> 82         self._locations_boundaries_setter(locations, boundaries)
     83
     84         # Metadata

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in _locations_boundaries_setter(self, locations, boundaries)
    228             if boundaries is None:
    229                 locations.sort()
--> 230                 self._check_bounds(locations)
    231                 self._locations = locations
    232                 # reset all the boundaries to True!!

/Users/jairideout/dev/scikit-bio/skbio/metadata/_interval.py in _check_bounds(self, locations)
    252                              'larger than upper bound (%r) or smaller than '
    253                              'lower bound (%r).' %
--> 254                              (locations, upper_bound, lower_bound))
    255
    256     @property

ValueError: Cannot set `locations` ([(0, 5)]) with coordinate larger than upper bound (4) or smaller than lower bound (0).

In [11]:
```

jairideout replied:
I understand the source of my confusion. `IntervalMetadata` doesn't explicitly define the upper bounds of `locations` or `upper_bound` itself to be inclusive or exclusive, and from the API/usage of the object it isn't clear either. I understood upper bounds to be inclusive for two reasons:

1) The constructor parameter `upper_bound` is documented as:

```
upper_bound : int
        Defines the upper bound of the interval features. No coordinate can
        be greater than it.
```

"No coordinate can be greater than it" implies (to me at least) that a coordinate equal to the upper bound is acceptable.

2) An interval's range(s) are specified as `locations`. `locations` implies coordinates/positions in the coordinate space, so it wasn't clear that the upper bound of a location follows Python's interpretation of slices and ranges.

I understand now that upper bounds are *always* exclusive, and lower bounds are always inclusive, in order to match Python's notion of slices/ranges. There isn't an off-by-one error given this interpretation of bounds.

I can see this potentially causing confusion with users (for example, I've played with the object and reviewed the code, and had an entirely different interpretation of the upper bound). @ebolyen and I talked about this and think it is important at this stage to make the following changes to `IntervalMetadata`/`Interval` in a separate PR in order to make this clearer:

1) Change all occurrences of `locations` to `bounds`. `locations` implies specific positions in the coordinate space, while the name `bounds` is more akin to Python's notion of slices and ranges, which has lower and (exclusive) upper bounds.

2) Change all occurrences of `boundaries` to `fuzzy`. `boundaries` could be misinterpreted to mean inclusive/exclusive for the bounds in `locations`, but it means something entirely different (unsure where the start and/or end really is). "fuzzy" matches Biopython's nomenclature too for their equivalent data structure. "fuzzy" is also a distinctive name that won't get confused with defining inclusive/exclusive bounds, and most users will see the name and ignore it (which covers most use-cases). When `fuzzy` is used, the booleans should be inverted, such that `True` means a lower/upper bound is "fuzzy", and `False` means we have a non-fuzzy (precise) understanding of the lower/upper bound.

3) With inclusive lower bound and exclusive upper bound, an interval with the same start and end (e.g. `[1, 1)`) doesn't represent a position in the coordinate space. These types of intervals should be disallowed to avoid potential confusion (and it isn't clear what exactly it means in the context of interval metadata). If a user wants an interval spanning a single position, use `[1, 2)`.

4) Document `bounds` to always be inclusive lower and exclusive upper. Document `upper_bound` in a similar way.

I think these changes will help make the object more easily understood by users. A couple of my comments on this PR don't apply given the interpretation of upper bounds, I'll follow up on those now.

RNAer replied:
I can imagine the confusion given current doc and naming. will change it as you request here.

jairideout replied:
Great, thanks @RNAer!

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+            with self.assertRaisesRegex(ValueError, '\(9\).*\(%d\)' % i):
+                self._interval_metadata_constructor_(i, self.im)
+
+    def test_constructor_no_interval_metadata(self):
```
You have tests for missing interval metadata but no tests for interval metadata that is present. Can you add the following test cases?

- test that a length-1 object can have interval metadata set thru the constructor (e.g. similar to `test_constructor_with_positional_metadata_len_one`). This is the base case (i.e. the smallest object that can have interval metadata)
- test that a length-greater-than-1 object can have interval metadata set thru the constructor (e.g. similar to `test_constructor_with_positional_metadata_len_greater_than_one`)

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+            self.assertFalse(obj.has_interval_metadata())
+            self.assertIsInstance(obj.interval_metadata, IntervalMetadata)
+
+    def test_constructor_makes_shallow_copy_of_interval_metadata(self):
```
Can you add a test case similar to `test_constructor_handles_missing_positional_metadata_efficiently`?

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+        obj1 = self._interval_metadata_constructor_(self.upper_bound)
+        obj2 = self._interval_metadata_constructor_(self.upper_bound)
+        self.assertReallyEqual(obj1, obj2)
+
```
add a test similar to `test_eq_handles_missing_positional_metadata_efficiently`

RNAer replied:
this test doesn't make sense to me for either interval or positional metadata mixins, because even the `_init_` implementation uses inefficient deleter, this test will still pass. If it is for testing of empty interval metadata, this is tested in other tests. So I think thies is a wast code.

RNAer replied:
I added this test anyway, in case you think otherwise. We can delete them in a future PR if I am correct.

jairideout replied:
This test is important because it asserts `__eq__` doesn't turn an internal `None` representation of interval metadata into an empty `IntervalMetadata` object. The test also makes sense for `metadata` and `positional_metadata`. If the `__eq__` method had been implemented like you had it before (i.e. without the `has_interval_metadata` guards), this test would fail. Basically we don't want to turn `None` into an empty `IntervalMetadata` object for efficiency. If you don't use `has_interval_metadata` to guard that and just use the getter, it will turn `None` into an empty `IntervalMetadata` object.

This has a very real impact on performance when creating many of these objects (e.g. reading a FASTQ file into Sequence objects). We tried removing the `None` optimization and there was a huge degradation in performance. Thus, it is important to test it here and in the other places I requested.

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+
+        self.assertReallyNotEqual(obj1, obj2)
+
+    def test_copy_interval_metadata_empty(self):
```
add more assertions similar to those in `test_copy_positional_metadata_empty`

jairideout replied:
also add a test case similar to `test_copy_positional_metadata_none`

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+        self.im = IntervalMetadata(self.upper_bound)
+        self.intvls = [
+            {'locations': [(0, 1), (2, 9)], 'metadata': {'gene': 'sagA'}},
+            # repeat it
```
outdated comment?

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+        self.assertIsNot(obj, obj_copy)
+
+    def test_copy_interval_metadata(self):
+        self.im.add(**self.intvls[0])
```
can you use `self.intvls[1]` instead so that the mutable list value in `.metadata` can be asserted to be the same object? that's where the behavior differs between shallow vs. deep copy

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+            self.assertIsNot(i.metadata['gene'], j.metadata['gene'])
+            self.assertIs(i.metadata['product'], j.metadata['product'])
+
+    def test_deepcopy_interval_metadata_empty(self):
```
add more assertions similar to those in `test_deepcopy_positional_metadata_empty`

jairideout replied:
also add a test case similar to `test_deepcopy_positional_metadata_none`

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+        obj.interval_metadata.add(**self.intvls[1])
+        self.im.add(**self.intvls[1])
+        self.assertEqual(obj.interval_metadata, self.im)
+
```
Can you add a test similar to `test_positional_metadata_getter_no_positional_metadata`?

jairideout replied:
Also add a test case that asserts the setter works with valid IntervalMetadata. It's being implicitly tested in some of these other test cases but it's good to have an explicit test for this, in case the other tests change in the future. e.g. similar to `test_positional_metadata_setter_data_frame`

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+                obj.interval_metadata = i
+
+        self.assertEqual(self.im, obj.interval_metadata)
+
```
Can you add a test case similar to `test_positional_metadata_setter_len_mismatch`?

jairideout commented on skbio/metadata/_testing.py:
```diff
@@ -904,3 +905,245 @@ def test_has_positional_metadata(self):
+        obj = self._interval_metadata_constructor_(self.upper_bound, self.im)
+        self.assertTrue(obj.has_interval_metadata())
+
+    def test_reverse(self):
```
Sorry for requesting this test, it's unnecessary b/c we're already asserting that the IntervalMetadata object is constructed correctly in several different situations, and that the object's axis length is respected in the IntervalMetadata object. We're not testing specific methods on IntervalMetadata here, and `reverse` isn't a special case. Feel free to remove this test.

jairideout commented on skbio/metadata/tests/test_mixin.py:
```diff
@@ -76,5 +79,40 @@ def __deepcopy__(self, memo):
+class TestIntervalMetadataMixin(unittest.TestCase, ReallyEqualMixin,
+                                IntervalMetadataMixinTests):
+    def setUp(self):
+        super()._set_up()
```
This is a little confusing -- can you just define a `setUp` method in `IntervalMetadataMixinTests`, and call `super().setUp` here? That's usually how I've seen this handled.

RNAer replied:
I view it the opposite way: note that `IntervalMetadataMixinTests` is not inherited from `TestCase`, so  naming it `setUp` may cause more confusion. More importantly, call `super().setUp` will call `TestCase.setUp()` rather than `IntervalMetadataMixinTests.setUp` if I name `_set_up` to `setUp`.

jairideout replied:
That makes sense, thanks for clarifying why it is called `_set_up`.

RNAer commented on skbio/metadata/_mixin.py:
```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata(
+                self._interval_metadata_axis_len_())
```
I disagree on that. We shouldn't make this an exceptional case. Plus, positional metadata can have 0 rows. interval metadata should behave similarly. And this should be perfectly legal code (although it is meaningless):
```
im = IntervalMetadata(0)
im.add([(0, 0)])  # Again, behaving like Python index, this feature has zero length.
```

jairideout replied:
My comment here no longer makes sense given the interpretation of exclusive upper bounds. It's great that an empty object (e.g. Sequence) can have empty positional and interval metadata!

I don't think we should allow start/end bounds that are the same (e.g. `(0, 0)`) because it doesn't represent any place in the coordinate space and users won't be able to query one of these intervals with `.query(locations=[(0,0)]`. Seems clearer to disallow these types of intervals entirely. It also matches how positional metadata works -- you can't have metadata attached to an "empty" position.

RNAer commented:
@jairideout , I replied on a few of your comments. Other than those, all are addressed. could you comment and take another pass? Thanks!

jairideout commented:
@RNAer, I replied to your comments and I think we've cleared up the confusion with inclusive vs. exclusive upper bounds. I made a comment about issuing a separate PR to make this clearer in `IntervalMetadata`/`Interval`, I'll review/merge that one when it's ready and we can come back to this PR (which is close to merge).

mortonjt commented:
👍 @jairideout what do you think?

jairideout commented:
Looks good, thanks for making those changes @RNAer!

RNAer commented:
yay!

On Tue, Sep 27, 2016 at 3:14 PM, Jai Ram Rideout <notifications@github.com>
wrote:

> Merged #1423 <https://github.com/biocore/scikit-bio/pull/1423>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1423#event-804627691>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA9CeNz1nFfdXF6rfHzim1aPIYgsugj1ks5quZVjgaJpZM4J2USg>
> .
>

mortonjt commented:
🍻 Thanks guys for getting this in!
